Nintendo switch support

KiritoDv opened the pull request:
Added complete nintendo switch support compilation

KiritoDv committed:
Moved gfx effects to the gpu and removed loadtexbyname on some textures

KiritoDv committed:
Readded empty if

Commit:
Added nintendo switch support

Commit:
Fixed compilation errors and window closing issue on linux

Commit:
Fixed kaleido on 64 bit

Commit:
Merge branch 'zelda64' into znx

Commit:
Merge branch 'zelda64' into znx

Commit:
Merge branch 'zelda64' into znx

Commit:
Fixed audio on nintendo switch

Commit:
Implemented performance mode switcher

Commit:
Added suggestions from pr review

KiritoDv committed:
Merge remote-tracking branch 'ship/develop' into znx

KiritoDv committed:
Fixed ftruncate errors

KiritoDv committed:
Fixed compilation errors

KiritoDv committed:
Added missing include rule

KiritoDv committed:
Fixed libstorm path

KiritoDv committed:
Removed ftruncate extern to use a correct posix version

guaycuru committed:
Merge branch 'develop' from HarbourMasters into znx (#3)

* Add Support for macOS (#441)

* Fixed soh filters

* add more makefile changes

* almost ready

* more updates

* update

* update

* Update Makefiles to handle both platforms

* Allow for overriding the CXX and CC executables

* Restore original structure while supporting custom CXX flags

* Remove some platform specific libs

* Dynamic target name

* Make X11 paths package-agnostic

* Remove changes to `gfx_opengl.cpp`

* Use OpenGL2 on MacOS instead of OpenGL3

* make it actually render something

* render at least the first texture, still need to figure out the second
one

* Let’s use OpenGL 3 again

* maybe this works to get the right texture? link's eyes still look off a bit

* did this work?

* set the platform to macos

* actual numbers are right, but logic is ugly XXX/TODO, i know

* add zlib to ldflags for ZAPDUtils

* A bit of cleanup

* Revert unneeded changes

* Remove GL_CHECK

* Fix issues with z64 branch

* use an std::map instead of a giant array

* three point filter fix (#2)

* Fix mac compilation

* fix audio for 64 bit

* revert audio heap size, keep bigger pools

* Add more Apple specific checks to our modifications

* Add building instructions for macOS

* Remove unecessary step from building instructions

* Add missing SDL2 & GLEW to Linux LDLIBS

* Update BUILDING.md

Co-authored-by: BountyChocolate123456 <101743444+BountyChocolate123456@users.noreply.github.com>

* Update soh/.gitignore to include other arch binaries

Co-authored-by: BountyChocolate123456 <101743444+BountyChocolate123456@users.noreply.github.com>

* Use right platform name for debugging window

Co-authored-by: BountyChocolate123456 <101743444+BountyChocolate123456@users.noreply.github.com>

* Fix stormlib on macos (arm64)

* Simplify some of the ifdef checks

* Revert an older no longer necessary fix

* Remove remaining unecessary deviations

* Update building instructions after StormLib changes

* Feature: Use OpenGL 4.1 (#1)

* Further tweak the BUILDING

* Tidy up

* reword -j message

* Add Jenkins CI Support (#2)

* Fix type issues

* add target <appbundle> and <filledappbundle>

add makefile targets to create an .app
`filledappbundle` creates the target with the .otr included

this should perhaps be moved to Application Support though

* pull gcc's rpath from otool output

* move make target to the end so it's not default

* Add Jenkins and make exe in par with other platforms

* Actually save build artefacts

* Fix artefact path

* Remove x11 mentions and linking (not used)

* Update building instructions for generating app

* use appsupport directory

* Add new app icon

* Update target to match macOS types

* Update more audio types

* fix null deref in Audio_PlayFanfare

* Remove old import from z64

* address final nit with apple ifdefs

Co-authored-by: KiritoDev <36680385+KiritoDv@users.noreply.github.com>
Co-authored-by: Jeffrey Crowell <github@crowell.biz>
Co-authored-by: BountyChocolate123456 <101743444+BountyChocolate123456@users.noreply.github.com>

* Move macOS to parallel stage

* fix linux crash

* Ignore all named saves

Necessary after new save format

* DPad items

* Move UI location arrays from static to global to accomodate customizable UI

* FIX: Increased the maximum actor limit for draw distance

Fix issue #529

* FIX: Remove duplicates of gNewDrops from the enhancements menu

* Get rid of erroneous int casts

* casts jya_cobra's shadowTexture pointer to uintptr_t instead of s32.

* Fix ability to assign equipment you don't own

* Fix title cards for market potion shop and bombchu shop

* Invalidate message item icon texture before drawing

* Add margins for dpad

* Fixed some instances where the original OOT save is erroneously preserved or recreated, leading to another conversion next time you launched the game

* Headphones setting in game should now work.
Resolves #13

* Map Select name fixes

* Change Power to Upgrades

* Rename Soh* files

* Update names

* Add more files

* Add more files

* Add toggles for disabling heart drops and random drops (#507)

* Adds toggles for disabling random drops and fixed heart drops

* Changed the "No Fixed Heart Drops" toggle to the more useful "No Random Heart Drops" toggle

* Adds an explanation that the no heart drops option is similar to Hero Mode

* Moved difficulty options into their own sub-menu
Moved time saver options into their own sub-menu
Moved clutter reducing options into their own sub-menu

* fixed the tag name of mweep speed

* Moved Skip Text to Experimental

* Fix testing masks of DPad

* Fix audio, boot commands, and save staes.

* add menu from https://github.com/PurpleHato/Shipwright/tree/ResetMenu

* implement reset

* Update README.md

* command on mac

* special char doesn't work with default imgui font

* change name based on PR comment

* missed a couple spots

* change sunset in save editor so dampe is out

* Update to latest `docking` imgui (#582)

* Update to latest `docking` imgui

commit: 67410d53f739b6a0df138e2252f0e5136b42062f

* Fix modified import for SDL

* Update imgui_impl_sdl.cpp

* Add Imgui include path for better imports

* Controller Navigation Fix-Up (#544)

* Controller Navigation Fix-Up

* [MISC] Fix typo

* Use Macro for __FILE__ & __LINE__ when possible (#559)

* First batch some overlay

* Almost all overlay

* effect & gamestate

* kaleido stuffs

* more overlay

* more left over from code folder

* remaining hardcoded line and file

* Open & Close _DISP __FILE__ & __LINE__ clean up

* Some if (1) {} remove

* LOG_xxxx __FILE__ , __LINE__ cleaned

* ASSERT macro __FILE__ __LINE__

* mtx without line/file in functions

* " if (1) {} " & "if (0) {}" and tab/white place

* LogUtils as macro

* GameState_, GameAlloc_, SystemArena_ & ZeldaArena_

* Revert "GameState_, GameAlloc_, SystemArena_ & ZeldaArena_"

This reverts commit 0d85caaf7e342648c01a15fe21e93637352dc596.

* Like last commit but as macro

* Fix matrix not using macros

* use function not macro

* DebugArena_* functions
GameAlloc_MallocDebug
BgCheck_PosErrorCheck as macros
removed issues with ; in macro file

* MM Bunny Hood enhancements: no age restrictions & wear across entrances (#560)

* Remove age restriction on MM Bunny Hood

* Handle mask slot changing; init Link wearing mask

* Reset mask state after selling

* Use unordered maps in some hot paths (#566)

* Use unordered maps in some hot paths

* Address PR comments

* Cheat: Infinite Epona Boost (#577)

* Cheat: Infinite Epona Boost

* Edit for readability

* update pulseaudio for 44.1khz audio (#587)

* use pulseaudio defaults

* spaces/tabs

* pulseaudio values tuned for n64 audio at 44.1khz

* update tlength

* remove one hardcoded samplerate

Co-authored-by: briaguya <briaguya@alice>

* Restoration Item Sliders (#595)

* Allows players to set custom restoration values for potions and milk

* Fixed a dumb oversight

* implemented fairy sliders

* use ini window resolution (#410)

* use ini window resolution

* use arguments for directx window size

* default 640x480 not 320x240

* kick off a build

* and revert

* default config 640x480 not 320x240

* add todo

Co-authored-by: briaguya <briaguya@alice>

* 3D Drops & Projectiles update (#548)

* Fix remaining rupee and keys

* fix tabs -> spaces

* more fix about tabs -> spaces

* Better up <-> down mouvement for hearts

* remove useless comment

* Cosmetics menu rework (#589)

* -

* DPad/some margin fix(left side)

* various ImGui stuff and fixes

* disabled kaleido menu (non working)

* fix win being dumb

* same for C btn this time

* Fix windows build

* Beating hearts fix

* Default win size and some placement fix

* Fix Dpad Ammo & C notes, Kaleido, white dog

* some texts fix and tab move

* Add stone of agony, some fixes and build correctly

* precise item place with Dpad

* Gamecube -> GameCube

Co-authored-by: Baoulettes <iMacWin10>

* [CI] Create an Appimage for 64bit builds (#570)

* [CI] Create AppImage file

* [CI] Updated Jenkinsfile

* [MISC] fix indentation

* [MISC] Fix indentation

* [MISC] Fix indentation

* [CI] 7z appimage and readme

* [CI] AppImage icon file (#597)

* [CI] Add patchelf

* [BUILD] Include Hi-Res icon

* Fix portability use of std::clamp (#596)

* Add N64 weird frames and OOB Bombchus cvars (#602)

* Add gN64WeirdFrames and gBombchuOOB cvars

* Rename gBombchuOOB to gBombchusOOB

* Fix pipeline

* [BUILD] Do not strip ZAPD (#598)

* [BUILD] Do not strip ZAPD

* [BUILD] Reuse Icon Asset

* [APPIMAGE] Handle paths with spaces

* [macos] Distribute optimized builds (#599)

* Cosmetics texts fixes (#604)

* branch init + few fixes

* Reset button

* Several texts fixes and default color to GameCube

* space issues

* some more extra space that was not needed.

* space yes last time ...

* Arrows being able to be rainbow some more texts fixes

* more texts fixes

* c-buttons some plurial fix

* Fixes bug in demo effect due to removed display list set.

* King dodongo room crash (#613)

* first attempt

* force camera setting to be at least 0 (dirty!)

* -

* reverted something to just include fix

* Invalidate text box icon before drawing (#607)

* [APPIMAGE] Set lib path for ZAPD (#611)

* New drop runtime fix (#614)

* inital rework

* fixed default rotation for 2D sprites

* fix tab/space issues

* Fix Ganon's Castle title cards (#606)

* Checkered room crash workaround (#615)

* Increase poly and vertex count for dyna actors

* Add TODO

* Only apply fix in Forest Temple

* Enable globally

* Match case of 'Shipwright' in Linux build (#619)

* Fix 0xabababab crash (#617)

* Unconditionally setup the normal skybox

* Only call Skybox_Setup once ever

* Equipment upgrade text fix and any pause slot enhancement fix (#620)

* Fix upgrade name text not rendering

* Use cvar and don't render equip help for empty item slots

* Fix rendering logic

* Fix incorrect item name rendered for one frame

* Reorder comparison

* Remove extra indent

* Remove accidental changes

* ImGui grammer, consistency, and clarity fixes (#625)

* ImGui grammer, consistency, and clarity fixes

Co-Authored-By: Stormghetti <56653191+Stormghetti@users.noreply.github.com>

* Adds more new lines for users on smaller displays

Co-Authored-By: Stormghetti <56653191+Stormghetti@users.noreply.github.com>

Co-authored-by: Stormghetti <56653191+Stormghetti@users.noreply.github.com>

* New drops shadows & rotation fixes (#627)

* fixes

* forgot to add heart pieces in rotation logic

* Dodongo's Cavern blue warp crash fix (#622)

* Doodong's Cavern blue warp crash fix

* >= not >

* Don't waste a line of space

* add support for clang compiler (#592)

* hacks to align strings for clang... wow just wow

* start work to getting built with clang

* fix issues with struct constructors, all builds, doesn't link still

* fix some narrowing issues that clang complains about

* fix compliation of zapd

* fix null deref in VersionInfo

* builds with clang

* make stringbuilding use StringHelper instead of addition

* fix linking

* add CLANG SHIP overlay on clang built versions

* doesn't need to be volatile

* mark unknown strings as extern

* rename some stuff

* can't align extern

* hopefully fix compilation for everythign

* expandtab

* allow setting LD

* Revert "allow setting LD"

This reverts commit 711aba6db2c41bab476bd34e878af6a37a7f5559.

maybe to use lld it should be a LDFLAG?

* -Wno-deprecated-declarations is required for newer versions of clang

on macOS 13 beta sdk, the version of apple clang requires this

* Add jenkins support for clang

* Forward CXX flags to stormlib compilation

* Move GCC only flags to check

* use exports to set multiarch setup

* Fix Jenkins forever

* use make instead of cmake --build

add some flags to build with clang-11 as well

* address review coments

- rework extraction to allow multi thread
- misc readability cleanup

* update makefile to add WARN on linux+clang

Co-authored-by: David Chavez <david@dcvz.io>

* Fix develop

* Fixes grey screen issue + tooltip for 2 handed shield

* Don't close controller after SDL has quit (#642)

* Don't close controller after SDL has quit

* Don't check if controller can rumble if null

* Fixes Barinade's set whenever boss fight is reloaded (#639)

* Reimplements unused Barinade reset function.

* Removes the Jellyfish from Barinade's body on reset.

* add the ability to pull graves during the day (#637)

* change max internal res multiplier to 3x, use float and % instead of int (#638)

* change max internal res multiplier to 3x, use float and % instead of int

* Update libultraship/libultraship/ImGuiImpl.cpp

Co-authored-by: Kenix3 <kenixwhisperwind@gmail.com>

Co-authored-by: briaguya <briaguya@alice>
Co-authored-by: Kenix3 <kenixwhisperwind@gmail.com>

* Fixes crash on Biggoron trade failure screen wipe

Resolves #621

* fixes center docking (#652)

* testing out item replacement (#416)

* skip learning song of storms

* don't set flag when getting goron tunic as child

* Initiates prelude check when master sword unloads.

Not quite how N64 rando does it but so far it's the only way I've found to make it trigger without also triggering the time travel again.

* Stops Shadow Temple lore prompts from appearing in rando.

* Skips cutscene of royal tomb explosion in rando.

Explosion sound doesn't play correctly and I think the debris appears in the wrong place, but the functionality is here.

* Improves visual of exploding gravestone.

* Adds some comments explaining the rando differences

* Skip ruto text box in jabu blue warp

For rando

* skip intro cutscene in dodongo's cavern

* load spoiler files on boot, fix spoilerfile existing check when making new saves

* name entry dropped spoiler logic

* make sure to actually init the cvar

* no chime on load

* uncomment

* Skip ganondrof cutscene

Skip to scream part of the death animation, skipping the text boxes etc. For rando

* Update z_boss_ganondrof.c

* skip owl flight cutscenes in rando

* Fixes skipped text so it only applies to shadow temple.

Earlier fix inadvertently applied to some other text as well, changed logic so that only specified sceneNums and textIds can have this enabled, and text skipped by sceneNum can have the skip overriden by textId if needed. Currently there are no overrides so the textId section of the logic is commented out to avoid compilation errors.

* Adds a default to the switch case statements that leaves the randoSkipText variable unchanged, just in case.

* TEST: Text for item

* Adding ganon flavor text

* ADD: AMMO Count

* format ganon text/hint text

* Autoskip the tower cutscene if settings call for tower collapse.

* ganon hint text logic

* Improved prelude after time travel fix

* swapped the sizes between ganon hint text and ganon text, as they were set to the wrong things.

* this is all i did

* not the cleanest code ever but it's working

* ADD: GS Count

* ADD: Wallter (crash for now)

* TWEAK: Wallet check

* FIX: Use DrawItem instread of DrawUpgrade... b-baka!

* Fixes some vanilla bugs introduced by rando code.

* Added cutscene skip for zelda escaping

Using the debug cutscene skipping function. Also added a conditional so the bridge doesn't spawn closed when cutscene is ready to trigger

* ADD: X Spacing + Placeholders for song

* ADD: default case for items

* TWEAK: Spacing

* FIX: Light Arrow

* ADD: Ammo Option

* use groups instead

* ADD: More spacing logic

* songs and names

* TWEAK: Color on wallet

* colors

* Added flags cutscene before nabooru fight

* ADD: ChromaKey text

* First attempt skip cs after nabooru defeat

* Better implementation for specific rando cutscene skips

* use pulseaudio defaults

* spaces/tabs

* move color push/pop to stop crash

* make the colors work again

* the real bottle fix

* pulseaudio values tuned for n64 audio at 44.1khz

* update tlength

* remove one hardcoded samplerate

* Cleaned up and fixed zelda escape skip

The if statement is a freaking monster, but unless we want to skip more cutscenes in the same way later, this is the most compact way of doing it that I know of.

* Revert one line to match original

nothing functional

* another hint line that breaks autonewline logic

* don't autospawn epona if we don't have the song/ocarina

* Trying to use iron knuckle death effects

not working yet

* Streamlined OoT cutscene skip for future additions

Also cleaned up if statement in general

* Made if statement more readable

Also added clarity for what cutscene was skipped

* Fixed typo in comment

* Janky nabooru defeat cs skip

* altar text formatting (gonna need help shortening some of the french ones)

* more altar text formatting

* english altar text formatting complete

* make gtg blocking guard check for card not bridge

* FIX: Typo!

* FIX: Uppercases

* FIX: Typo

* TWEAK: Alter + some names

* TWEAK: More caps!

* ADD: Missing string

TWEAK more uppercases and namefixe
s

* Hide nabooru death by covering her in flames

* bandaid fix for death crash issue

* Twinrova defeat cs skip

Skips the animation and manually calls the function to show the "beam" around the sisters

* fix crash

* fix caps to match

* fix great fairy reward mashing/shielding issue

* TWEAK : Typo clé to Clé

* TWEAK: Some Altar hints

TWEAK: Some capitals

* TWEAK: Unmatching text + some cap again

* TWEAK: More tweaks

* fix build

* remove extra json.hpp, add hint

* Update randomizer_item_tracker.cpp

* TWEAK: Double Defense with RedParticles instead of white

* make sure we don't optimize out the check to ensure a spoilerfile exists

* vanilla ganon boss key hint formatting

* TWEAK: FR- better way of the hero text

* fix

* and again

* Initializes dungeonsDone items in gSaveContext to 0.

* Replaces sizeof calculation with a NUM_DUNGEONS constant.

* Fixes Saria's Gift on the LW Bridge from getting killed when holding shield.

* More airtight fix for Saria's Gift on the Bridge.

* Lifts one of the conditions in the if statement a little higher to prevent unnecessary lookups of getItemId.

* Invalidate text box icon before drawing

* Fixes the case where Saria's gift is an Ice Trap.

We still get the Ice Trap once, but never again. This does mean you can now hold R while walking in to avoid the ice trap, but everything else seems to work fine.

* Initial commit

Might need changing when we change the settings in the future

* Fixes Door of Time opening cutscene after warping with prelude.

* Initial waterfall skip

Very rudimentary way of doing things but it seems to work so :shrug:

* inital rework

* fixed default rotation for 2D sprites

* fix tab/space issues

* 3d drops rando merge fix again

* Allows Impa to appear in the Lullaby check post drawbridge escape.

* Changes Ganon's Trials Count setting to a checkbox

The checkbox is whether or not to skip all of them. Leaving the box unchecked will mean doing all of them. Eventually this will be switched back to a slider once we implement the logic for which trials start out completed.

* Sets all Ganon's Trials to incomplete in new saves.

Fixes https://github.com/briaguya-ai/rando-issue-tracker/issues/131

* fix castle guards when oot throw cutscene has already played in rando

* Properly removes the beams when trials are cleared.

* Removes Question Mark from Skip Ganon's Trials UI.

* Adds a todo comment about when to change back to slider.

* make deku seeds check for bullet bag

* Various tweaks

TWEAK: Altar Text
TWEAK: Hint names
TWEAK: Replace more problematic œ to oe

* upgrade ocarina on both child and adult equips

* FIX: Jabu Item

* update equipped hookshot/longshot when obtained as other age

* add hint

* don't give the bgs check without the claim check

* Skips Darunia Cutscene in Fire Temple

* Added a TODO note about not skipping the cutscene.

There is a setting we will want to have eventually that will require this cutscene to not be skipped since it is used during a glitch.

* remove todo

* restore fast ocarina option in imgui that was lost in merge

* Fixes grey screen issue + tooltip for 2 handed shield

* update to use dg instead of g for textures in item tracker

* TWEAK: Default color for cosmetic RAND button was not the corect one

* fix texture crash, remove unused item tracker code

* don't open mask shop until we get zelda's letter

* Update README.md

* Prevents "correct" chime under incorrect conditions.

* Fixes typo in conditional and adds "bonk" sound effect.

"Bonk" sound is NA_SE_SY_OCARINA_ERROR and it plays when conditions for the Door of Time have not been met after playing Song of Time. This is only possible in rando's "Intended" Door of Time option, in which the Ocarina of Time and all 3 spritual stones are required to open the door, instead of the vanilla requirements of just having the song of time.

* remove modify dpad equips toggle, replace with checks for dpad menu

* remove extra check

* add ability to hold c-up to assign to dpad when dpad menuing is enabled

* disable d-pad navigation on item menu when holding c-up to equip

* dpad+c-up stuff for equipment menu

* ADD: Checbox for songs colors

* TWEAK: RandoColors for normal songs

* kind of quick and dirty but it works

* TWEAK: Clarity of the tooltip

Co-authored-by: briaguya <briaguya@alice>
Co-authored-by: Christopher Leggett <chris@leggett.dev>
Co-authored-by: aMannus <mannusmenting@gmail.com>
Co-authored-by: PurpleHato <linkvssangoku.jr@gmail.com>
Co-authored-by: Dog <5172592+Dog@users.noreply.github.com>
Co-authored-by: Vague Rant <vaguerant@users.noreply.github.com>
Co-authored-by: Baoulettes <perlouzerie@hotmail.fr>
Co-authored-by: Ada <60364512+GreatArgorath@users.noreply.github.com>

* Cosmetics hotfixes (#640)

* Initial branch creation

* Revert Main Game so it do not conflict later

* should fix window build, made namespace for Cosmetics

* forgot to edit one title

* Ability to add hidden window (usefull for Rainbow)
Fix building issues

* Line break, unused bool remove

* add descriptive todo for death crash bandaid (#655)

Co-authored-by: briaguya <briaguya@alice>

* Removed legacy audio mode and fixed ganon sound bug (#657)

* Free Camera (#337)

* wip free cam

* Almost done, needs collision still

* Added free cam behind cvar

* added WIP collision

* Fixed & implemented "Manual mode" from WW & TP

* Fixed camera not rotating when Link is moving

* fixed initialized camera rotation

* Fixed camera getting stuck + made it smoother

* reduced deadzone

* fixed epona camera height + added WW z-target free camera

* Adjusted player camera height & fixed fov

* Fixed camera roll

* fixed fov when moving the camera while in z-target

* Camera resets to Auto when going through doors or changing maps

* Fixed building

* touch

* more touch work

* Added WIP mouse support to the free cam

* gui stuff

* fixed building

* fixed building error

* ok fixed building for real this time

* oops

* Fix compilation issues

* removed mouse stuff that magically appeared in this branch

* smoothed out stick values & removed remains of mouse support

* re-added manual camera when pressing Z

* reduced minimum Y position of camera

* Addressed dcsv's nitpicks

* part 2

* oops

Co-authored-by: David Chavez <david@dcvz.io>

* Rando: Allows Malon's Item Check to be obtained by pulling out the Ocarina. [FIXED PR] (#672)

* Fixes using the Ocarina to get the check from Malon.

Still some cleanup to do here. For some reason the player can shield before receiving the check. It doesn't set the flag if the player does that so they can still try again, but would prefer a different solution if possible.

* Prevents Shielding from blocking the Item_Give from happening.

* Code Cleanup and comments explaining the new rando flow.

* Removes inventory check when pulling out Ocarina

This allows OI to properly give the check, which is important for Glitched logic later down the line. Talking to Malon still requires the Ocarina in your inventory.

* Prevents non-malon textboxes from triggering the check.

Also adds a comment explaining the condtional for getting the check from talking to Malon since it got pretty long.

* Actually fixes checking for text boxes.

* Relocates a comment for improved clarity.

* Fix Rando Water Temple Softlock (#665)

We use 3DS logic to generate item placement, but didn't have this specific door in Water Temple unlocked from the beginning like 3DS does.

This meant that if people took specific paths through the temple, they could softlock themselves by missing a key.

* Rando: GtG and carpenter prompts skip (#663)

* Skip gtg and carpenter prompt

For rando. Tested and just works.

* Fixed missing break

* Hide debug overlay behind gDebugEnabled (#660)

* Introduce App Directory Path (#572)

* Introduce app directory path concept

* macos: Remove hacky way of using applicaiton directory

* Update the new SaveManager

* Address stack user after return

* Remove unecessary property

* Use std::string for filepath

* Improve clang specific detections

* Use new path system for imgui files

* Improve helper for getting relative paths

* fix hidden wnd (#744)

* Split damage multiplication into its own PR (#656)

* Split damage multiplication into its own PR

* Found a more elegant implementation of the powers char*[]

* Fixes Maps, Compasses, and Boss Keys in Vanilla. (#751)

* .xiF slebaL

* Update Keese labels

* Fixed soundfont issues

* Skip warp song cutscenes in rando (#664)

Does it by skipping to the last part of the cutscene data. Tested on all songs, both adult and child.

* don't spawn blocking mido after we've already shown him the sword/shield (#675)

Co-authored-by: briaguya <briaguya@alice>

* Controller Configuration UI and JSON Config (#760)

* Initial controller hud ui

* Reverted fbdemo changes

* Moved config to json and implemented controller config

* fix build on linux, gitignore new config file

* fix build

* Fix compilation and file directory paths

* Call save on cvar save

* Fixed cvar loading and added deck slots to the config

* Changed control deck port 0 to use a physical device by default

* Added gyro and rumble & fixed loading errors

* Save config on toggle menubar

* fix linux build

* Fixed drift calculation

* Controller config now saves when pressing F1

* Removed ExitGame hook from ImGuiImpl

* Moved mappings to a map

* Added GetKeyName

* untranslate scancodes

* Fixed hud layout on keyboard device

* Fixed keyboard read on hud

* Fixed crash when reloading controllers

* Removed ConfigFile and changed file extension

* Changed Dummy to Disconnected and fixed filters

* Removed function leftover

* Changed ControllerHud to InputEditor

Co-authored-by: briaguya <briaguya@alice>
Co-authored-by: David Chavez <david@dcvz.io>

* Enough! My ship sails in the morning.

* Fixed menubar items position (#763)

* Fixed menubar items position

* Reverted tooltip tab position

* Fixes macOS randomizer functionality with App Directory (#761)

* Fixes macOS randomizer functionality with App Directory

* Fix windows build

* Update soh/soh/Enhancements/randomizer/3drando/rando_main.cpp

* Update soh/soh/Enhancements/randomizer/3drando/spoiler_log.cpp

* Revert band-aid fix

Co-authored-by: Kenix3 <kenixwhisperwind@gmail.com>

* Fix migration cvar deletion path (#765)

* Various controller fixes (#771)

* Fix controller

* Also fix rumble strength being a bool

* Remove ControllerHud.cpp

* Downgrade platform toolset back to previous version

* Fix gyro

* Fix bug that makes binding axes difficult and clear buttons before reading

* Exaggerate gyro display and adjust stick binding threshold

* Initialize drift thresholds

* git subrepo push soh

subrepo:
  subdir:   "soh"
  merged:   "75ccbade8"
upstream:
  origin:   "https://github.com/HarbourMasters/soh.git"
  branch:   "master"
  commit:   "75ccbade8"
git-subrepo:
  version:  "0.4.1"
  origin:   "???"
  commit:   "???"

* One more change from PR review

* Fix some paths

* Fix merge conflict messup

* More merge conflict fixes

* And another conflict fix

* And another fix

* Remove reference to removed build files

* Add full path to switch cmake

Co-authored-by: David Chavez <david@dcvz.io>
Co-authored-by: KiritoDev <36680385+KiritoDv@users.noreply.github.com>
Co-authored-by: Jeffrey Crowell <github@crowell.biz>
Co-authored-by: BountyChocolate123456 <101743444+BountyChocolate123456@users.noreply.github.com>
Co-authored-by: sholdee <102821812+sholdee@users.noreply.github.com>
Co-authored-by: briaguya <briaguya@alice>
Co-authored-by: rozlette <Rozelette@users.noreply.github.com>
Co-authored-by: PurpleHato <linkvssangoku.jr@gmail.com>
Co-authored-by: Christopher Leggett <chris@leggett.dev>
Co-authored-by: Kenix3 <kenixwhisperwind@gmail.com>
Co-authored-by: vaguerant <vaguerant@users.noreply.github.com>
Co-authored-by: earthcrafterman <banddstudios@gmail.com>
Co-authored-by: louist103 <35883445+louist103@users.noreply.github.com>
Co-authored-by: briaguya <70942617+briaguya-ai@users.noreply.github.com>
Co-authored-by: qurious-pixel <62252937+qurious-pixel@users.noreply.github.com>
Co-authored-by: Baoulettes <perlouzerie@hotmail.fr>
Co-authored-by: Sirius902 <10891979+Sirius902@users.noreply.github.com>
Co-authored-by: modestposer <modestposer@gmail.com>
Co-authored-by: Ada <60364512+GreatArgorath@users.noreply.github.com>
Co-authored-by: Stormghetti <56653191+Stormghetti@users.noreply.github.com>
Co-authored-by: Sirius902 <3645979-Sirius902@users.noreply.gitlab.com>
Co-authored-by: MelonSpeedruns <melonspeedruns@outlook.com>
Co-authored-by: aMannus <mannusmenting@gmail.com>
Co-authored-by: Dog <5172592+Dog@users.noreply.github.com>
Co-authored-by: Nicholas Estelami <NEstelami@users.noreply.github.com>
Co-authored-by: ChristopherJTrent <ChristopherJTrent@outlook.com>
Co-authored-by: agamache <aaroncgamache@gmail.com>
Co-authored-by: M4xw <m4x@m4xw.net>

KiritoDv committed:
Fixed switch compilation

KiritoDv committed:
Fixed responsive on switch imgui

KiritoDv committed:
Remove config save hook

KiritoDv committed:
Added physical slots into controller name

guaycuru committed:
Add `ifdef` for switch specific code

guaycuru committed:
Add `ifdef` for switch specific code

guaycuru committed:
Merge branch 'HM_develop' into znx

KiritoDv committed:
Added applet mode detection

KiritoDv committed:
Replaced homebrew icon

KiritoDv committed:
Fixed randomness on applet mode view

KiritoDv committed:
Merge branch 'develop' into znx

guaycuru committed:
Merge branch 'kiritodv_znx' into znx_merge

dcvz commented on StormLib/CMakeLists.txt:
```diff
@@ -282,7 +289,7 @@ set(TEST_SRC_FILES
 )
 
-add_definitions(-D_7ZIP_ST -DBZ_STRICT_ANSI)
+add_definitions(-D_7ZIP_ST -DBZ_STRICT_ANSI -D_POSIX_C_SOURCE=200809L)
```
we don't have to add this to all builds- you can check:

```
if(NOT WIN32 AND NOT APPLE AND NOT ${CMAKE_SYSTEM_NAME} STREQUAL FreeBSD)
  # Enable POSIX extensions such as `readlink` and `ftruncate`.
  add_definitions(-D_POSIX_C_SOURCE=200809L)
endif()
```

dcvz replied:
this is changing the definitions for all platforms. Let's separately add this:
```cmake
if(NOT WIN32 AND NOT APPLE AND NOT ${CMAKE_SYSTEM_NAME} STREQUAL FreeBSD)
  # Enable POSIX extensions such as `readlink` and `ftruncate`.
  add_definitions(-D_POSIX_C_SOURCE=200809L)
endif()
```

KiritoDv committed:
Fixed clkrst initialization

KiritoDv committed:
Switch profiles cleanup

dcvz commented on StormLib/CMakeLists.txt:
```diff
@@ -15,6 +15,13 @@ option(STORM_BUILD_TESTS
 #   "BUILD_TESTING" OFF # Stay coherent with CTest variables
 )
 
+if(DEFINED _initial_CMAKE_TOOLCHAIN_FILE
```
I don't think you need all this..

dcvz commented on libultraship/libultraship/Archive.cpp:
```diff
@@ -327,7 +327,11 @@ namespace Ship {
 #ifdef _WIN32
 		std::wstring wfullPath = std::filesystem::absolute(MainPath).wstring();
 #endif
+#if defined(__SWITCH__)
```
question: why do you have to change these? `std::filesystem:` is present in the switch

KiritoDv replied:
Because filesystem::absolute doesnt work on the switch port

dcvz commented on libultraship/libultraship/Cvar.cpp:
```diff
@@ -14,7 +14,7 @@ extern "C" CVar* CVar_Get(const char* name) {
 }
 
-extern "C" s32 CVar_GetS32(const char* name, s32 defaultValue) {
+extern "C" int CVar_GetS32(const char* name, int defaultValue) {
```
all of the changes to `int` in this file should be `int32_t` if you `#include <stdint.h>` in the header

dcvz commented on libultraship/libultraship/Cvar.h:
```diff
@@ -1,16 +1,14 @@
 #ifndef _CVAR_H
 #define _CVAR_H
 
-#include <PR/ultra64/gbi.h>
```
replace it with `#include <stdint.h>` and you can change the `int` to `int32_t`

dcvz commented on libultraship/libultraship/Environment.cpp:
```diff
@@ -9,6 +9,11 @@ namespace SohUtils {
         environmentVars[key] = value;
     }
+
+    bool hasEnvironmentVar(const std::string& key) {
```
This does not seem to get used at all

dcvz commented on libultraship/libultraship/Environment.h:
```diff
@@ -4,5 +4,6 @@
 
 namespace SohUtils {
     void saveEnvironmentVar(const std::string& key, const std::string& value);
+    bool hasEnvironmentVar(const std::string& key);
```
This does not seem to get used at all

dcvz commented on libultraship/libultraship/GlobalCtx2.cpp:
```diff
@@ -72,6 +72,8 @@ namespace Ship {
 #ifdef _WIN32
             MessageBox(nullptr, L"Main OTR file not found!", L"Uh oh", MB_OK);
+#elif defined(__SWITCH__)
+            printf("Main OTR file not found!\n");
```
Is SPDLog not working on the Switch or is it for nxlink? I wonder if its worth stubbing out calls to `SPDLOG` on the Switch and having it internally use `printf`.

Kenix3 replied:
We always call SPDLOG with a macro. It takes a different format than printf, so we'd have to stub out completely or remove logging on switch.

dcvz commented on libultraship/libultraship/ImGuiImpl.cpp:
```diff
@@ -26,6 +26,7 @@
 #include "Lib/Fast3D/gfx_rendering_api.h"
 #include "Lib/spdlog/include/spdlog/common.h"
 #include "Utils/StringHelper.h"
+#include "SwitchImpl.h"
```
```suggestion
#ifdef __SWITCH__
#include "SwitchImpl.h"
#endif
```

dcvz commented on libultraship/Makefile:
```diff
@@ -49,6 +50,11 @@ CPPFLAGS := -MMD
 
 MMFLAGS := -Wno-deprecated-declarations -ObjC++ -fobjc-weak -fobjc-arc
 
+ifeq ($(X11), 1)
```
this seems like an unnecessary change.. do you need this for something?

dcvz commented on libultraship/libultraship/Cvar.cpp:
```diff
@@ -5,7 +5,7 @@
 #include <utility>
 #include <PR/ultra64/gbi.h>
-#include "imgui_internal.h"
+#include "Lib/ImGui/imgui_internal.h"
```
why did this header change? looks like you need to add "Lib/ImGui" to the includes in `lus/Makefile.switch`

dcvz commented on libultraship/libultraship/ImGuiImpl.cpp:
```diff
@@ -717,8 +728,13 @@ namespace SohImGui {
         if (ImGui::BeginMenuBar()) {
             if (DefaultAssets.contains("Game_Icon")) {
+            #ifdef __SWITCH__
+                ImVec2 iconSize = ImVec2(20.0f, 20.0f);
```
is it too much if we just make it 20 for all platforms?

KiritoDv replied:
Yes, because we dont need the resize on other platorms, imgui renders at 2.0 scale on switch

dcvz commented on libultraship/libultraship/InputEditor.cpp:
```diff
@@ -173,44 +185,52 @@ namespace Ship {
-			}
-			ImGui::SetCursorPosX(cursorX);
-			DrawVirtualStick("##GyroPreview", ImVec2(-10.0f * Backend->wGyroY, 10.0f * Backend->wGyroX));
+				float cursorX = ImGui::GetCursorPosX() + 5;
```
the margins for all these lines got messed up

dcvz commented on libultraship/libultraship/InputEditor.cpp:
```diff
@@ -173,44 +185,52 @@ namespace Ship {
-			ImGui::InputFloat("##GDriftY", &profile.Thresholds[DRIFT_Y], 1.0f, 0.0f, "%.1f");
-			ImGui::PopItemWidth();
-			ImGui::EndChild();
+				ImGui::SameLine();
```
the margins for all these lines got messed up

dcvz commented on libultraship/libultraship/InputEditor.cpp:
```diff
@@ -115,15 +115,23 @@ namespace Ship {
 			DrawButton("Z", BTN_Z);
 			DrawButton("START", BTN_START);
 			SEPARATION();
+	#ifdef __SWITCH__
```
there's so much custom layouting for the switch.. I wonder if there's a better way to describe these UI's.

dcvz commented on libultraship/libultraship/Lib/Fast3D/gfx_glx.cpp:
```diff
@@ -1,4 +1,4 @@
-#ifdef __linux__
+#if defined(__linux__) && defined(X11_SUPPORTED)
```
so why is this check required? The Switch is not using GLX..

dcvz commented on libultraship/libultraship/Lib/Fast3D/gfx_opengl.cpp:
```diff
@@ -31,10 +31,15 @@
 #else
 #include <SDL2/SDL.h>
 #include <GL/glew.h>
+#ifndef __SWITCH__
```
you don't need this check.. we'll never enter this #else statement.

dcvz commented on libultraship/libultraship/Lib/Fast3D/gfx_glx.cpp:
```diff
@@ -156,6 +156,7 @@ static struct {
     Atom atom_wm_delete_window;
 
     bool is_fullscreen;
+    bool is_running = true;
```
I guess you're trying to fix something on linux here? seems a bit off topic for the PR.

dcvz commented on libultraship/libultraship/Lib/Fast3D/gfx_opengl.cpp:
```diff
@@ -471,6 +476,11 @@ static struct ShaderProgram* gfx_opengl_create_and_load_new_shader(uint64_t shad
         append_line(fs_buf, &fs_len, "texel.rgb = mix(texel.rgb, new_texel, vGrayscaleColor.a);");
     }
 
+    if(cc_features.opt_grayscale) {
```
is this a fix for something? we're changing the grayscale shader..  also the if right above this already checks for `opt_grayscale`.. do these need to be combined?

dcvz commented on libultraship/libultraship/Lib/Fast3D/gfx_sdl2.cpp:
```diff
@@ -15,13 +15,18 @@
 #endif
 
 #include "../../ImGuiImpl.h"
+#include "../../SwitchImpl.h"
```
let's wrap this in a `#if __SWITCH__`

dcvz commented on libultraship/libultraship/Lib/ImGui/backends/imgui_impl_sdl.cpp:
```diff
@@ -757,6 +771,7 @@ static void ImGui_ImplSDL2_CreateWindow(ImGuiViewport* viewport)
         SDL_GL_MakeCurrent(vd->Window, backup_context);
 
     viewport->PlatformHandle = (void*)vd->Window;
+#if defined(_WIN32) || defined(__APPLE__)
```
what about Linux? isn't it better to do `#if !defined(__SWITCH)`?

dcvz commented on libultraship/libultraship/Lib/ImGui/backends/imgui_impl_sdl.cpp:
```diff
@@ -411,6 +413,7 @@ static bool ImGui_ImplSDL2_Init(SDL_Window* window, SDL_Renderer* renderer, void
     // Our mouse update function expect PlatformHandle to be filled for the main viewport
     ImGuiViewport* main_viewport = ImGui::GetMainViewport();
     main_viewport->PlatformHandle = (void*)window;
+#if defined(_WIN32) || defined(__APPLE__)
```
what about Linux? isn't it better to do `#if !defined(__SWITCH)`?

dcvz commented on libultraship/libultraship/Lib/ImGui/backends/imgui_impl_sdl.cpp:
```diff
@@ -74,8 +74,10 @@
 #include <SDL_syswm.h>
 #else
 #include <SDL2/SDL.h>
+#ifndef __SWITCH__
```
does this cause issues? this header is available in switch sdl I think

dcvz commented on libultraship/libultraship/Lib/ImGui/backends/imgui_impl_sdl.cpp:
```diff
@@ -654,6 +659,15 @@ static void ImGui_ImplSDL2_UpdateMonitors()
     }
+#else
+        int width, height;
+        Ship::Switch::GetDisplaySize(&width, &height);
```
I think its great that we're adding Switch support to ImGui, maybe you even want to make a proper PR against the upstream version.. but I don't think its a great that ImGui is calling LUS here.. 

Let's just import `switch.h` and properly get what we need here.

dcvz commented on libultraship/libultraship/Lib/ImGui/backends/imgui_impl_sdl.cpp:
```diff
@@ -63,9 +63,9 @@
 //  2016-10-15: Misc: Added a void* user_data parameter to Clipboard function handlers.
 
-#include "imgui.h"
+#include "../imgui.h"
```
it's not necessary to change these.. just add the missing include directory into your Makefile.

dcvz commented on libultraship/libultraship/Lib/ImGui/backends/imgui_impl_opengl3.h:
```diff
@@ -19,7 +19,7 @@
 
 #pragma once
-#include "imgui.h"      // IMGUI_IMPL_API
+#include "../imgui.h"      // IMGUI_IMPL_API
```
it's not necessary to change these.. just add the missing include directory into your Makefile.

dcvz commented on libultraship/libultraship/Lib/ImGui/backends/imgui_impl_opengl3.cpp:
```diff
@@ -91,7 +91,7 @@
 #endif
 
-#include "imgui.h"
+#include "../imgui.h"
```
it's not necessary to change these.. just add the missing include directory into your Makefile.

dcvz commented on libultraship/libultraship/Lib/ImGui/backends/imgui_impl_opengl3.cpp:
```diff
@@ -243,7 +246,7 @@ bool    ImGui_ImplOpenGL3_Init(const char* glsl_version)
 
     // Initialize our loader
-#if !defined(IMGUI_IMPL_OPENGL_ES2) && !defined(IMGUI_IMPL_OPENGL_ES3) && !defined(IMGUI_IMPL_OPENGL_LOADER_CUSTOM)
+#if !defined(__SWITCH__) && !defined(IMGUI_IMPL_OPENGL_ES2) && !defined(IMGUI_IMPL_OPENGL_ES3) && !defined(IMGUI_IMPL_OPENGL_LOADER_CUSTOM)
```
this is not the right way to check here. ImGui provides the option of setting `IMGUI_IMPL_OPENGL_LOADER_CUSTOM` -- you can set that above where you import glad and then you don't need any checks here.

dcvz commented on libultraship/libultraship/Lib/ImGui/backends/imgui_impl_opengl3.cpp:
```diff
@@ -132,7 +132,10 @@
+#elif defined(__SWITCH__)
+#define GL_GLEXT_PROTOTYPES 1
+#include "glad/glad.h"
+#else
```
Let's restore this back.. we're losing logic here.
```suggestion
#elif !defined(IMGUI_IMPL_OPENGL_LOADER_CUSTOM)
```

dcvz commented on libultraship/libultraship/Lib/ImGui/backends/imgui_impl_opengl3.cpp:
```diff
@@ -132,7 +132,10 @@
-#elif !defined(IMGUI_IMPL_OPENGL_LOADER_CUSTOM)
+#elif defined(__SWITCH__)
+#define GL_GLEXT_PROTOTYPES 1
+#include "glad/glad.h"
```
we're importing a lib here.

```suggestion
#include <glad/glad.h>
```

dcvz commented on libultraship/libultraship/Lib/Fast3D/gfx_opengl.cpp:
```diff
@@ -31,10 +31,15 @@
 #include <GL/glew.h>
+#elif __SWITCH__
+#include <SDL2/SDL.h>
+#include "glad/glad.h"
```
we're importing a lib here
```suggestion
#include <glad/glad.h>
```

dcvz commented on libultraship/libultraship/Lib/Fast3D/gfx_sdl2.cpp:
```diff
@@ -15,13 +15,18 @@
+#elif __SWITCH__
+#include <SDL2/SDL.h>
+#include <switch.h>
+#include "glad/glad.h"
```
we're importing a lib here 
```suggestion
#include <glad/glad.h>
```

dcvz commented on libultraship/libultraship/Lib/ImGui/backends/imgui_impl_sdl.cpp:
```diff
@@ -63,9 +63,9 @@
+#include "../imgui.h"
 #include "imgui_impl_sdl.h"
-
+#include "../../../SwitchImpl.h"
```
I don't think we should connect ImGui to LUS..

dcvz commented on libultraship/libultraship/Lib/ImGui/backends/imgui_impl_sdl.h:
```diff
@@ -18,7 +18,7 @@
 
 #pragma once
-#include "imgui.h"      // IMGUI_IMPL_API
+#include "../imgui.h"      // IMGUI_IMPL_API
```
it's not necessary to change these.. just add the missing include directory into your Makefile.

dcvz commented on libultraship/libultraship/SwitchImpl.cpp:
```diff
@@ -0,0 +1,272 @@
+        swkbdConfigMakePresetDefault(&kbd);
+}
+
+// TODO: Fully Implement Keyboard
```
this is still a todo.. did you want to address this before the PR is merged?

dcvz commented on libultraship/libultraship/Lib/ImGui/imgui_widgets.cpp:
```diff
@@ -4344,6 +4344,7 @@ bool ImGui::InputTextEx(const char* label, const char* hint, char* buf, int buf_
         const bool is_paste = ((is_shortcut_key && IsKeyPressed(ImGuiKey_V)) || (is_shift_key_only && IsKeyPressed(ImGuiKey_Insert))) && !is_readonly;
         const bool is_undo  = ((is_shortcut_key && IsKeyPressed(ImGuiKey_Z)) && !is_readonly && is_undoable);
         const bool is_redo  = ((is_shortcut_key && IsKeyPressed(ImGuiKey_Y)) || (is_osx_shift_shortcut && IsKeyPressed(ImGuiKey_Z))) && !is_readonly && is_undoable;
+        const bool is_overwrite = state->OverwriteData != NULL;
```
I'm a bit wary of adding new functionality into ImGui.. is this not achievable via the official API?

dcvz commented on libultraship/libultraship/Lib/Mercury/Mercury.cpp:
```diff
@@ -1,11 +1,13 @@
 #include "Mercury.h"
 
+#include "Hooks.h"
```
does not seem like these changes are needed?

dcvz commented on soh/soh/Enhancements/bootcommands.c:
```diff
@@ -20,7 +20,7 @@ extern BootCommandFunc BootCommands_Command_LoadFileSelect(char** argv, s32 argc
                                    { "--loadfileselect", BootCommands_Command_LoadFileSelect } };
 
-void BootCommands_Init()
+void BootCommands_Init() 
```
```suggestion
void BootCommands_Init()
```

dcvz commented on soh/soh/Enhancements/bootcommands.c:
```diff
@@ -64,8 +64,8 @@ void BootCommands_ParseBootArgs(s32 argc, char** argv)
 
-    for (i = 0; i < argc; i++)
-        DebugArena_Free(argv[i]);
+    //for (i = 0; i < argc; i++)
```
dos this cause issues? perhaps we should disable it only for switch devices if its really necssary

dcvz commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -44,6 +44,7 @@
 #include "variables.h"
 #include "macros.h"
 #include <Utils/StringHelper.h>
+#include "SwitchImpl.h"
```
Let's wrap this in a __SWITCH__ check

dcvz commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -251,9 +256,10 @@ extern "C" void Graph_ProcessGfxCommands(Gfx* commands) {
+
                 int samples_left = AudioPlayer_Buffered();
                 u32 num_audio_samples = samples_left < AudioPlayer_GetDesiredBuffered() ? SAMPLES_HIGH : SAMPLES_LOW;
-                // printf("Audio samples: %d %u\n", samples_left, num_audio_samples);
```
doesn't seem like this had to go

Kenix3 replied:
Also we don't use printf

dcvz commented on soh/src/code/main.c:
```diff
@@ -42,7 +42,7 @@ void main(int argc, char** argv)
     InitOTR();
     BootCommands_Init();
 
-    BootCommands_ParseBootArgs(argc - 1, (char**)&argv[1]);
```
do these need to be commented out? if so, shall we add a __SWITCH__ check?

dcvz reviewed (commented):
@KiritoDv here's my first pass review!

KiritoDv committed:
Cleaned code based on dcvz comments

dcvz committed:
Remove unused hooks in Mercury

KiritoDv committed:
Fixed compilation

dcvz committed:
Use IMGUI_IMPL_OPENGL_LOADER_CUSTOM

dcvz committed:
Fix up StormLib modifications

dcvz committed:
Merge branch 'develop-zhora' into znx_merge

# Conflicts:
#	libultraship/libultraship/ImGuiImpl.cpp
#	soh/soh/OTRGlobals.cpp

dcvz committed:
Handle touch events at ImGui/SDL level

dcvz committed:
Enable opening menu with -

dcvz committed:
Load Nintendo Switch font

dcvz committed:
Updates to window initiation

dcvz committed:
Handle virtual keyboard via SDL

dcvz committed:
Print OTR missing message to screen

dcvz committed:
Rename makefile for Switch

dcvz committed:
Some more additional fixes

Baoulettes committed:
Branch creation plus edit

Changed ItemWidth & made label invisible (else issues)
Added a submit button next to the text field

KiritoDv committed:
Removed set_texture_filter on imgui draw

KiritoDv committed:
Readded nintendo switch font

KiritoDv committed:
Fixed randomizer crash

dcvz committed:
Fix issue with strdup windows

Kenix3 commented on libultraship/libultraship/Archive.cpp:
```diff
@@ -68,7 +68,7 @@ namespace Ship {
 
 		if (!attempt) {
-			printf("({%i}) Failed to open file {%s} from mpq archive {%s}", GetLastError(), filePath.c_str(), MainPath.c_str());
+			printf("(%i) Failed to open file { %s } from mpq archive { %s }\n", GetLastError(), filePath.c_str(), MainPath.c_str());
```
We don't use printf in libultraship. Look for uses of SPDLOG_ macros to understand how we log in LUS.
https://github.com/gabime/spdlog

Kenix3 commented on libultraship/libultraship/SwitchImpl.cpp:
```diff
@@ -0,0 +1,209 @@
+}
+
+const char* RandomTexts[] = {
+    "Psst, don't forget to blame melon",
```
We should probably have a conversation about some of these on Discord.

dcvz replied:
I think the idea here was to use rando's random texts right?

Kenix3 reviewed (changes requested):
Just a few small things. Please also see @dcvz's comments.

qurious-pixel committed:
[SWITCH] Update Jenkins file (#9)

* [SWITCH] Update Jenkins file

* [SWITCH] Dockerfile.switch

Can be combined with other Dockerfile

* [SWITCH] Add entrypoint file

* Update Jenkinsfile

* [SWITCH] Dockerfile mtab link 

link `/proc/self/mounts /etc/mtab`

Co-authored-by: David Chavez <david@dcvz.io>

KiritoDv committed:
Cleaned code based on PR comments

KiritoDv committed:
Fixed switch compilation

KiritoDv committed:
Cleaned full texture cache instead of doing it per char

KiritoDv committed:
Randomizer only saves when its not a vanilla save

KiritoDv committed:
Disabled input when menubar is opened

dcvz commented on libultraship/libultraship/SwitchImpl.cpp:
```diff
@@ -0,0 +1,225 @@
+}
+
+void Ship::Switch::SetupFont(ImFontAtlas* fonts) {
+    return;
```
I think this was an accident again @KiritoDv ?

KiritoDv replied:
D: Yes

KiritoDv committed:
Removed return from SetupFont

KiritoDv committed:
Merge branch 'ship/develop-zhora' into znx_merge

GaryOderNichts commented on ZAPDTR/ZAPDUtils/Utils/Directory.h:
```diff
@@ -4,6 +4,10 @@
 #include <vector>
 
+#ifdef __SWITCH__
+long int pathconf (const char *__path, int __name){}
```
Not a fan of this approach. This can quickly lead to multiple definitions when included in multiple files.
The Wii U port places this in a seperate file: <https://github.com/GaryOderNichts/Shipwright/blob/wiiu/soh/wiiu/pathconf.c>

GaryOderNichts commented on libultraship/Makefile.switch:
```diff
@@ -0,0 +1,191 @@
+		libultraship/Lib/stb/stb_impl.c \
+
+SOURCEFILES_CPP :=	\
+	$(shell find libultraship/Factories -name "*.cpp") \
```
You can add these directories to the `SOURCES` variable to avoid extra shell commands in here.

GaryOderNichts commented on libultraship/Makefile.switch:
```diff
@@ -0,0 +1,191 @@
+
+CFLAGS	:=	-g -Wall -ffunction-sections \
+			$(ARCH) $(DEFINES) \
+            -DSPDLOG_NO_THREAD_ID -DSPDLOG_NO_TLS \
```
The switch should support TLS afaik?

GaryOderNichts commented on libultraship/Makefile.switch:
```diff
@@ -0,0 +1,191 @@
+            -DSPDLOG_NO_THREAD_ID -DSPDLOG_NO_TLS \
+		    -DSTBI_NO_THREAD_LOCALS -DDEBUG
+
+CFLAGS	+=	$(INCLUDE) -D__SWITCH__ -DENABLE_OPENGL -DSDL_BYTEORDER
```
Why the `SDL_BYTEORDER` define?

GaryOderNichts commented on libultraship/libultraship/SDLController.cpp:
```diff
@@ -441,7 +450,7 @@ namespace Ship {
 
     const char* SDLController::GetControllerName() {
-        return SDL_GameControllerNameForIndex(physicalSlot);
+        return strdup(ControllerName.c_str());
```
Is the `strdup` necessary?
The pointer returned by `c_str` is valid as long as `ConrollerName` is not destroyed.

KiritoDv replied:
Yes

GaryOderNichts commented on soh/Makefile.switch:
```diff
@@ -0,0 +1,280 @@
+# source files
+#-------------------------------------------------------------------------------
+SOURCEFILES_C := \
+	$(shell find soh -type f -name "*.c") \
```
See comment above regarding `SOURCES`.

GaryOderNichts commented on libultraship/libultraship/Window.cpp:
```diff
@@ -252,6 +254,10 @@ namespace Ship {
             ControllerApi->SaveControllerSettings();
         });
+
+        ModInternal::RegisterHook<ModInternal::RequestConfigSave>([]() {
```
Why is that hook necessary? Doesn't seem to be called anywhere.

KiritoDv committed:
Cleaned code based on comments

KiritoDv committed:
Cleaned up random texts on switch error screen

Kenix3 commented on libultraship/libultraship/SwitchImpl.cpp:
```diff
@@ -0,0 +1,286 @@
+   }
+}
+
+const char* RandomTexts[] = {
```
These are the approved messages:
```c
    "Psst, don't forget to blame Melon",
    "Potsanity when?",
    "Why are you acting so random?",
    "Enough! My ship sails in the morning",
    "Do you want 2 or 7 of those?",
    "Lamp oil, rope, bombs you want it, it's yours my friend as long as you have enough rupees",
    "You can build it yourself",
    "Descargar para android",
    "Made with <3 by the Harbour Masters!",
    "They say that Kenix is not a developer",
    "Squadala we're off",
    "They say one once saw an equals not get set equals",
    "This is the port all true gamers dock at"
    "Enhancements? Times Savers? Cheats? You want them? They're yours my friend!",
    "They say you gotta have the BIIIIG salad",
    "They say Louis stopped working on the imports so he can focus on the exports",
    "They say ZAPD is good software",
```

dcvz commented:
LGTM 👍
